### PR TITLE
feat: add a keyed subscription, unsusbcription, and status APIs

### PR DESCRIPTION
**Pull request by icehaunter**

### CODE REVIEW

File: clients/typescript/src/satellite/shapes/shapeManager.ts [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+  serverId?: string
+  overshadowsFullKeys: string[]
+  shapes: Shape[]
+  shapeHash: string
+  fullKey: string
+}
+
+type OptionalRecord<T> = Record<string, T | undefined>
+
+export class ShapeManager {
```
Comment (msfstef) [inline clients/typescript/src/satellite/shapes/shapeManager.ts:21]: for the sake of introspecting the shape manager's internals, could we have a `listAllSubscriptions` API that returns all requested, unfulfilled, and active subscriptions in the form of
```ts
{ shape: Shape[], key: string, status: SyncStatus }[]
```

where the key is either the user provided one or the shape hash.

Comment (msfstef) [reply clients/typescript/src/satellite/shapes/shapeManager.ts:21]: On second thought I can add that after this PR along with a notifier API as I'd need both

**Review by msfstef — APPROVED** [review]
Looks good, awesome work!

I've left some comments regarding making some things more clear or perhaps adding a couple more tests

File: clients/typescript/src/satellite/shapes/shapeManager.ts [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+function onlyDefined<T>(x: T | undefined): x is T {
+  return x !== undefined
+}
+
+function makeFullKey(hash: string, key: string): string {
+  return hash + ':' + key
+}
+
+function splitFullKey(fullKey: string): [hash: string, key: string] {
+  return splitOnce(fullKey, ':')
```
[the remark this answers is not in the document]
Comment (icehaunter) [reply clients/typescript/src/satellite/shapes/shapeManager.ts:377]: Yup, as specified in a comment where we join the key

File: clients/typescript/src/satellite/shapes/manager.ts [hunk 3]
```diff
@@ -20,18 +19,15 @@ export type GarbageCollectShapeHandler = (
...
+export class InMemorySubscriptionsManager implements SubscriptionsManager {
   private inFlight: SubcriptionShapeRequests = {}
   protected fulfilledSubscriptions: SubcriptionShapeDefinitions = {}
   private readonly shapeRequestHashmap: Map<string, SubscriptionId> = new Map()
 
   private readonly gcHandler?: GarbageCollectShapeHandler
 
   constructor(gcHandler?: GarbageCollectShapeHandler) {
-    super()
+    // super()
```
[the remark this answers is not in the document]
Comment (icehaunter) [reply clients/typescript/src/satellite/shapes/manager.ts:30]: Actually this entire class is not used

File: clients/typescript/src/satellite/client.ts [hunk 4]
```diff
@@ -935,18 +935,18 @@ export class SatelliteClient implements Client {
   }
 
   private handleTransaction(message: SatOpLog) {
-    if (!this.subscriptionsDataCache.isDelivering()) {
-      this.processOpLogMessage(message)
-    } else if (this.inbound.receivingUnsubsBatch) {
+    if (this.inbound.receivingUnsubsBatch) {
```
Comment (msfstef) [inline clients/typescript/src/satellite/client.ts:938]: need to check if this is an empty array (or set it to undefined/false)

**Pull request merged**